### PR TITLE
GPU memory model

### PR DESCRIPTION
**Pull request by Naraenda**
Closes https://github.com/RadeonOpenCompute/ROCm/issues/2265

### CODE REVIEW

**Review by saadrahim — CHANGES_REQUESTED** [review]
Need to add this to the toc.yml as well

**Comment by saadrahim** [comment]
Also, please take a look at https://docs.olcf.ornl.gov/systems/crusher_quick_start_guide.html#enabling-gpu-page-migration

**Review by afanfa — CHANGES_REQUESTED** [review]

File: docs/understand/gpu_memory.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+If the application needs to use data on both host and device regularly, doesn't want to deal with separate allocations, and is not worried about maxing out the VRAM on MI200 GPUs (64 GB per GCD), use managed memory.
+:::
+
+:::{tip}
+If managed memory performance is poor, check to see if managed memory is supported on your system and if page migration (XNACK) is enabled.
+:::
+
+## Access behavior
+
+Memory allocations for GPUs that do not support XNACK or have it disabled by setting the environment variable `HSA_XNACK=0` behave as follow:
```
[the remark this answers is not in the document]
Comment (Naraenda) [reply docs/understand/gpu_memory.md]: I have removed the references to XNACK in this section. I want to have this section before explaining XNACK as not every GPU has the feature and it is not default behavior.

File: docs/understand/gpu_memory.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+| API                | Data location | Host access  | Device access |
+|--------------------|---------------|--------------|---------------|
+| System allocated   | Host          | Local access | Page fault    |
```
[the remark this answers is not in the document]
Comment (Naraenda) [reply docs/understand/gpu_memory.md]: Thanks! Is changing it to _unhandled page fault_ sufficient?

File: docs/understand/gpu_memory.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+| `hipMallocManaged` | Host          | Local access | Zero-copy     |
+| `hipHostMalloc`    | Host          | Local access | Zero-copy*    |
+| `hipMalloc`        | Device        | Zero-copy    | Local access  |
+
+Zero-copy access happen over the Infinity Fabric interconnect or PCI-E lanes on discrete GPUs.
+
+:::{note}
+While `hipHostMalloc` allocates memory on the device, kernels may not be able to use the pointer.
```
[the remark this answers is not in the document]
Comment (Naraenda) [reply docs/understand/gpu_memory.md]: My bad! I have changed it. Thanks!

File: docs/understand/gpu_memory.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+| `hipMalloc`        | Device        | Zero-copy                | Local access             |
+
+To check if page migration is available on a platform, use `rocminfo`:
+
+```sh
+$ rocminfo | grep xnack
+      Name:                    amdgcn-amd-amdhsa--gfx90a:sramecc+:xnack-
+```
+
+Here, `xnack-` means that XNACK is available but is disabled by default.
```
[the remark this answers is not in the document]
Comment (Naraenda) [reply docs/understand/gpu_memory.md:130]: Extra content has been added 😄

File: docs/understand/gpu_memory.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (afanfa) [reply docs/understand/gpu_memory.md]: yes

File: docs/understand/gpu_memory.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+Other managed memory API calls will have undefined behavior.
+It is therefore recommended to check for managed memory capability with: `hipDeviceGetAttribute` and `hipDeviceAttributeManagedMemory`.
+
+HIP supports additional calls that work with page migration:
+
+- `hipMemAdvise`
+- `hipMemPrefetchAsync`
+
+:::{tip}
+If the application needs to use data on both host and device regularly, doesn't want to deal with separate allocations, and is not worried about maxing out the VRAM on MI200 GPUs (64 GB per GCD), use managed memory.
```
Comment (afanfa) [inline docs/understand/gpu_memory.md:76]: A less informal sentence would be better. doesn't -> does not

File: docs/understand/gpu_memory.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+| `hipMallocManaged` | Host          | Local access | Zero-copy            |
+| `hipHostMalloc`    | Host          | Local access | Zero-copy*           |
+| `hipMalloc`        | Device        | Zero-copy    | Local access         |
+
+Zero-copy accesses happen over the Infinity Fabric interconnect or PCI-E lanes on discrete GPUs.
+
+:::{note}
+While `hipHostMalloc` allocated memory is accessible by a device, the host pointer must be converted to a device pointer with `hipHostGetDevicePointer`.
+
+Memory allocated through standard system allocators such as `malloc`, can accessed by device by registering the memory via `hipHostRegister`.
```
Comment (afanfa) [inline docs/understand/gpu_memory.md:99]: can accessed -> can be accessed by "a" device

Comment (afanfa) [reply docs/understand/gpu_memory.md:99]: This of course is not true for MI250X. A clarification would be good.

Comment (Naraenda) [reply docs/understand/gpu_memory.md:99]: > This of course is not true for MI250X. A clarification would be good.

I'm not super familiar with the specifics of the MI250X. Am I correct that you mean that using `hipHostRegister` on system allocated memory on the MI250X does not work? Does this also hold for the MI250 (non-X)?

From my limited reading, the MI250 and MI250X are both OAMs and the only difference is the amounts of CUs and less IF links.

Comment (afanfa) [reply docs/understand/gpu_memory.md:99]: "Memory allocated through standard system allocators such as `malloc`, can accessed by device by registering the memory via `hipHostRegister`." hipHostRegister works fine on MI250X, you can access host memory also via XNACK.

File: docs/understand/gpu_memory.md (hunk 5 above)
Comment (Naraenda) [reply docs/understand/gpu_memory.md]: Changed

File: docs/understand/gpu_memory.md (hunk 6 above)
Comment (Naraenda) [reply docs/understand/gpu_memory.md]: Is this better?

**Review by saadrahim — DISMISSED** [review]

**Review by saadrahim — DISMISSED** [review]

**Comment by Naraenda** [comment]
Brought in latest changes and resolved linting issues

**Review by saadrahim — APPROVED** [review]

**Pull request merged**